### DISCUSSION TITLE
Reading Kaifa MB310H4BNL

### DISCUSSION THREAD

**Discussion by rweijnen** [arendst/Tasmota#22354] [Q&A] [answered]
I'm trying to read data from a Kaifa MB310H4BNL in The Netherlands.
I am using this script:
```
>D
>B
=>sensor53 r
>M 1
+1,3,o,0,300,Kaifa,1,20,2F3F210D0A,063030300D0A
1,1.8.1(@1,E1,kWh,Tariff_1,1
1,1.8.2(@1,E2,kWh,Tariff_2,1
1,2.8.0(@1,E3,kWh,Total_export,1
1,2.8.1(@1,E4,kWh,Total_export,1
1,2.8.2(@1,E5,kWh,Total_export,1
#
```

When sending `/?!` (`2F3F210D0A`) meter responds with `/KFM5\2Kaifa MA309M`. Good this means we have communication!

KFM = Kaifa Metering
5 = baudrate 9600
2 = mode (send data continously without needing to resend /?!, not sure about this but read this in protocol specification)

I tried the script to switch baud mode but not getting valid output with it.

So continueing at 300 I do get output:
![image](https://github.com/user-attachments/assets/21ab2e5c-938b-43e7-9fb9-993958dff25e)

However I have 2 issues:

1. random control characters like <DLE> <BS> <FF> <STX> are in the output
2. the script seems to restart sending `/?!` before it has received all data.

As a result, the output is unreliable.

If I change `+1,3,o,0,300,Kaifa,1,20,2F3F210D0A,063030300D0A` to a longer value e.g. `+1,3,o,0,300,Kaifa,1,50,2F3F210D0A,063030300D0A` I don't get any output as it's taking to long to send the <ACK>

I also attempted to switch the baudrate with this script:
```
>D
res=0  
scnt=0    
;For this Example in the >F section  
>F
;count 100ms   
scnt+=1  
switch scnt  
case 6  
;set sml driver to 300 baud and send /?! as HEX to trigger the Meter   
res=sml(1 0 300)  
res=sml(1 1 "2F3F210D0A")  
;1800ms later \> Send ACK and ask for switching to 9600 baud  
case 18  
res=sml(1 1 "063035300D0A")  
;2000ms later \> Switching sml driver to 9600 baud    
case 20  
res=sml(1 0 9600)  ;Restart sequence after 50x100ms    
case 50  
; 5000ms later \> restart sequence    
scnt=0  
ends        
>M 1
+1,3,o,0,9600,Kaifa,1,600
1,1.8.1(@1,E1,kWh,Tariff_1,1
1,1.8.2(@1,E2,kWh,Tariff_2,1
1,2.8.0(@1,E3,kWh,Total_export,1
1,2.8.1(@1,E4,kWh,Total_export,1
1,2.8.2(@1,E5,kWh,Total_export,1
#
```

But then I get this output:
![image](https://github.com/user-attachments/assets/bfa84e3b-e0ea-4590-9bcb-807f201f7ded)

Any ideas or tips appreciated!

**Comment by rweijnen** [comment]
Not fully there yet but if I enter from the Web console: `SerialConfig 8E1`

Then by starting at 300 baud and switching to 9600 I can get output with `Sensor53 d1`

```
>D
res=0  
scnt=0    
;For this Example in the >F section  
>F
;count 100ms   
scnt+=1  
switch scnt  
case 6  
;set sml driver to 300 baud and send /?! as HEX to trigger the Meter   
res=sml(1 0 300)  
res=sml(1 1 "2F3F210D0A")  
;1800ms later \> Send ACK and ask for switching to 9600 baud  
case 18  
res=sml(1 1 "063035300D0A")  
;2000ms later \> Switching sml driver to 9600 baud    
case 20  
res=sml(1 0 9600)  ;Restart sequence after 50x100ms    
case 500  
; 5000ms later \> restart sequence    
scnt=0  
ends        
>M 1
+1,3,o8E1,0,9600,Kaifa,1,600
1,1.8.1(@1,E1,kWh,Tariff_1,1
1,1.8.2(@1,E2,kWh,Tariff_2,1
1,2.8.0(@1,E3,kWh,Total_export,1
1,2.8.1(@1,E4,kWh,Total_export,1
1,2.8.2(@1,E5,kWh,Total_export,1
#
```

Output:
![image](https://github.com/user-attachments/assets/944cb1e6-6ea5-4f70-977a-ab0c60d836f6)


Challenge is that my meter has 300:7E1 for the initial message and 9600:8E1 and I've not found a way to make Tasmota switch. If you look at the output above the first identification string is now a bit garbled

**Comment by rweijnen** [comment]
I'm making some progress but there's something that I don't understand. If I look at the examples in the [documentation](https://tasmota.github.io/docs/Smart-Meter-Interface/) it says:

> With a few meters, it is necessary to request the meter to send data using a specific character string. This string has to be sent at a very low baud rate (300Baud). If you reply the meter with an acknowledge and ask the it for a new baud rate of 9600 baud, the baud rate of the SML driver has to be change

The sample script is as follows:
```
>D
res=0  
scnt=0    
;For this Example in the >F section  
>F
;count 100ms   
scnt+=1  
switch scnt  
case 6  
;set sml driver to 300 baud and send /?! as HEX to trigger the Meter   
res=sml(1 0 300)  
res=sml(1 1 "2F3F210D0A")  
;1800ms later \> Send ACK and ask for switching to 9600 baud  
case 18  
res=sml(1 1 "063035300D0A")  
;2000ms later \> Switching sml driver to 9600 baud    
case 20  
res=sml(1 0 9600)   
;Restart sequence after 50x100ms    
case 50  
; 5000ms later \> restart sequence    
scnt=0  
ends        
>M 1
+1,3,o,0,9600, ,1  
;...etc. 
```

But this script lacks setting up the sensor which normally happens in BOOT:
``
>B
    =>sensor53 r
``

although some examples use:
```
>B
   -> sensor 5r r
```

Workaround:
Start with this script:
```
>D
>B
=>sensor53 r
>M 1
+1,3,o,0,300,Kaifa,1,20,2F3F210D0A,063030300D0A
1,=so2,2
1,1.8.0(@1,1.8.0 Energy In Total,kWh,Energy_In_T,1
1,1.8.1(@1,1.8.1 Energy In T1,kWh,Energy_In_T1,1
1,1.8.2(@1,1.8.2 Energy In T2,kWh,Energy_In_T2,1
1,2.8.0(@1,2.8.0 Energy Out Total,kWh,Energy_Out_T,1
1,2.8.1(@1,2.8.1 Energy Out T1,kWh,Energy_Out_T1,1
1,2.8.2(@1,2.8.2 Energy Out T2,kWh,Energy_Out_T2,1
#
````

Then change to this script:
```
>D
res=0
scnt=0
;For this Example in the >F section

>F
;count 100ms
scnt+=1
switch scnt
case 6
;set sml driver to 300 baud and send /?! as HEX to trigger the Meter
res=sml(1 0 300) 
res=sml(1 1 "2F3F210D0A")

;1800ms later \> Send ACK and ask for switching to 9600 baud
case 18
res=sml(1 1 "063035300D0A")

;2000ms later \> Switching sml driver to 9600 baud
case 20
res=sml(1 0 9600)

;Restart sequence after 50x100ms    
case 50  
; 5000ms later \> restart sequence    
scnt=0  
ends

>M 1
+1,3,o8E1,0,9600,Kaifa,1,600
1,=so2,2
1,1.8.0(@1,1.8.0 Energy In Total,kWh,Energy_In_T,1
1,1.8.1(@1,1.8.1 Energy In T1,kWh,Energy_In_T1,1
1,1.8.2(@1,1.8.2 Energy In T2,kWh,Energy_In_T2,1
1,2.8.0(@1,2.8.0 Energy Out Total,kWh,Energy_Out_T,1
1,2.8.1(@1,2.8.1 Energy Out T1,kWh,Energy_Out_T1,1
1,2.8.2(@1,2.8.2 Energy Out T2,kWh,Energy_Out_T2,1
#
```

the >M section is never read from the 2nd, it continues to use the one from 1st.

Questions:
What's the difference between `->` and `=>` in the >B section with Sensor53?

How to add `Sensor53 r` to the 2nd script (so I don't have to use the first) because if I add it I get a `duplicate GPIO defined` error.

**Comment by rweijnen** [comment, marked as answer]
Well, seems like I finally figured it out:
using `+1,3,o8E1,0,9600,Kaifa,1,600` is what's causing the pin conflict message (the `o8E1` part).

So set serialconfig: `SerialConfig 8E1` then this works for me:
```
>D
res=0
scnt=0
>B
->sensor53 r
;For this Example in the >F section

>F
;count 100ms
scnt+=1
switch scnt
case 6
;set sml driver to 300 baud and send /?! as HEX to trigger the Meter
res=sml(1 0 300) 
res=sml(1 1 "2F3F210D0A")

;1800ms later \> Send ACK and ask for switching to 9600 baud
case 18
res=sml(1 1 "063035300D0A")

;2000ms later \> Switching sml driver to 9600 baud
case 20
res=sml(1 0 9600)

;Restart sequence after 50x100ms    
case 250  
; 5000ms later \> restart sequence    
scnt=0  
ends

>M 1
+1,3,o,0,9600,Kaifa,1,600
1,=so2,2
1,1.8.0(@1,1.8.0 Energy In Total,kWh,Energy_In_T,1
1,1.8.1(@1,1.8.1 Energy In T1,kWh,Energy_In_T1,1
1,1.8.2(@1,1.8.2 Energy In T2,kWh,Energy_In_T2,1
1,2.8.0(@1,2.8.0 Energy Out Total,kWh,Energy_Out_T,1
1,2.8.1(@1,2.8.1 Energy Out T1,kWh,Energy_Out_T1,1
1,2.8.2(@1,2.8.2 Energy Out T2,kWh,Energy_Out_T2,1
#
```

Now going to check if it's stable when running longer time.

**Comment by bdoeksen** [comment]
@rweijnen Excellent work, thanks! Also for the integration in Home Assistant.
Does the meter not output current power as is shown on the display (the "P" value)? This would be useful for load balancing (SmartEVSE) or peak shaving.
Kind regards,
Bas.

**Reply by rweijnen** [reply]
I've collected all I've discovered on my [Github repo](https://github.com/rweijnen/Kaifa-MB310H4BNL) which perhaps you've seen.

Unfortunately in HDLC Mode C it does not return P, but I think it might be available in HDLC Mode E but my attempt to get that working failed:

```
Attempt to switch to mode E:
Initial communication at 300E1, then meter identifies as:
/KFM5\2Kaifa MA309M

Per protocol it supports 9600 baud and mode 2 (HDLC, mode E), if I send the sequences manually I get this:
-> /?! CR LF
<- /KFM5\2Kaifa MA309M
-> ACK 2 5 2 CR LF
<- ACK 252
```

but nothing after that :(

```
01:49:22 Initializing serial connection.
01:49:23 IEC Sending:/?!<CR><LF>
01:49:24 HDLC received: /KFM5\2Kaifa MA309M<CR><LF>
01:49:24 BaudRate is : 9600
01:49:24 Moving to mode E.
06 32 35 32 0D 0A
01:49:27 Send SNRM request.
7E A0 07 03 03 93 8C 11 7E
01:49:32 Data send failed. Try to resend 1/3
01:49:37 Data send failed. Try to resend 2/3
01:49:42 Failed to receive reply from the device in given time.
01:49:42 Disconnect request
01:49:42 Send Disconnect request.
7E A0 07 03 03 53 80 D7 7E
01:49:47 Data send failed. Try to resend 1/3
01:49:52 Data send failed. Try to resend 2/3
```

I've been thinking about capturing P from the digital display with an espcam but unfortunately the display keeps switching between the different readings. It's a shame that the grid operator, in my case Enexis, refuses to share any documentation!

**Reply by bdoeksen** [reply]
Yes, I've seen the repo. Well written!
A pity about the mode E.
My grid operator is Liander, but I don't expect they are willing to spend time on this either.
There are always CT clamps or kWh meters, but reading straight from the meter is just more... elegant.

**Reply by rweijnen** [reply]
That's very interesting info, I wasn't aware of other grid operators using this meter as well. So I'm curious if your meter returns exactly the same information as mine, or different/more.

Part of the information returned is firmware version, yours might be different.

Did you try asking Liander for documentation?

**Reply by bdoeksen** [reply]
No, I haven't tried that yet.

**Reply by rweijnen** [reply]
Hey @bdoeksen would be great if you could ask them, maybe there are more open than Enexis (I'm happy to draft an email for you if that helps)?

**Reply by bdoeksen** [reply]
Hi @rweijnen ! The use case for this is a little too obscure for me to put more effort into. Another factor to take into consideration: in order to make load balancing safe, we need values for current on all three phases. Just the value for net power is not enough to prevent main fuses from blowing out. I guess I'll just defect to the smart meter side and use the P1 interface. ;-)

**Reply by rweijnen** [reply]
I would appreciate if you could ask them anyway. Might help me and others and it's just asking them about documentation. If they do supply we can either use the knowledge or go back to Enexis and dispute their no

**Reply by bdoeksen** [reply]
I'd be happy to. I don't know whom at Liander to ask though. We'll have to find out.

**Reply by rweijnen** [reply]
I think you can use contact@liander.nl, they will probably need your meter number which you can find via [www.liander.nl/meternummer](https://www.liander.nl/meters-en-meterstanden/meternummer). Below a template in Dutch you can use or adapt:

Beste Klantenservice,

Ik heb een Kaifa MB310H4BNL digitale (niet slimme) meter. Deze meter beschikt over een S0 poort, niet te verwarren met de P1 poort van slimme meter (al lijkt het protocol er sterk op). Deze poort ondersteunt IEC 62053-21 protocol. 
Via een infrarood lezer lukt het om de actuele meterstanden (afname en teruglevering) uit te lezen, zoals hier op [github](https://github.com/rweijnen/Kaifa-MB310H4BNL) te vinden.

Maar ik kan het actuele verbruik (liefst per fase) niet uitlezen. Kunt u mij documentatie verstrekken hoe ik eea kan uitlezen en indien van toepassing pincode of andere benodigde authenticatiegegevens?

Mijn gegevens:
NAW
Meternummer: xxx

**Comment by Wets0cks** [comment]
@rweijnen , Thanks for information. As I understand you did get data from your Kaifa MB310H4BNL. I have a KAIFA MB310H4BNL5 and don't get any data what so ever. I understand that the "5" at the end means that it sends data every 1 second instead of every 5 seconds. I use a [Stromleser](https://stromleser.de/products/stromleser-tasmota) IR reading head with Tasmota and connected to Home Assistant with MQTT. All communication is fine, but no data. I tried several scripts, including yours, but still nothing. 
[Tasmota](https://tasmota.github.io/docs/Smart-Meter-Interface/#kaifa-mb310h4bde) and [Stromleser](https://stromleser.de/en/blogs/news/stromleser-wifi-und-ttl-lesekopf-einrichten-tasmota) keep saying that you need a PIN-code to enter the advanced settings menu in the Kaifa meter to set INF to "ON". I tried pressing the orange button 13 times, holding it for 10 seconds, but I cannot get it to enter advanced mode, if any exists. Can you tell me if you also needed to set the INF setting to "ON" or did your meter started sending data from the start?

![Main Menu](https://github.com/user-attachments/assets/f84c8949-2a4a-4d4d-9646-b7d400dff884)

**Reply by rweijnen** [reply]
I didn't have to press any buttons or mar any changes to my meter. I think such methods are only available in Germany or Austria. Most likely cause of failure is that your unit uses a different baudrate. In Tasmota you can enable mode where you can see all output. Did you try that?

**Comment by Wets0cks** [comment]
Do you mean Tasmota- configuration - logging -> set all loglevels to "Debug" or somewhere else?

**Comment by Wets0cks** [comment]
I simplified my script, set baudrate to 115200 but still no data. Any tips? 
<img width="1013" height="554" alt="script" src="https://github.com/user-attachments/assets/79afe693-0b11-4dbc-a45b-011d74fdee98" />
<img width="2581" height="1245" alt="Log" src="https://github.com/user-attachments/assets/c0226c41-3862-48b0-a066-306d8087b5a1" />

**Reply by rweijnen** [reply]
I think it's from the web ui console, command `sensor53 d<meter number>` so typically `sensor53 d1`.

Then start with 300 baud 7E1 and see if you have some output from `/?!`

**Reply by Wets0cks** [reply]
I now use your script that is 'marked as answer'. Next I send the command `SerialConfig 7E1` from the UI console followed by complete restart from the UI menu. Logging shows that the serial mode is changed to 7E1.
00:00:00.032 SER: Set to 7E1 115200 bit/s
00:00:00.032 SER: HWCDC supports 115200 bit/s only

But I just cannot get the baudrate down to 300 (or 9600), I tried:
10:36:12.734 CMD: sml(1 1 300)
10:36:12.743 MQT: stat/tasmota_B540A4/RESULT = {"Command":"Unknown","Input":"SML (1 1 300)"}
10:36:32.693 CMD: sml(1 1 "2F3F210D0A")
10:36:32.700 MQT: stat/tasmota_B540A4/RESULT = {"Command":"Unknown","Input":"SML (1 1 \"2F3F210D0A\")"}
How can I get the baudrate down to 300 or 9600 instead of 115200 ?

After that I tried this:
10:49:24.736 CMD: sensor53 d1
10:49:24.743 MQT: stat/tasmota_B540A4/RESULT = {"Time":"2025-11-03T10:49:24","SML":{"CMD":"dump: 1"}}
10:49:43.435 /?!
10:49:43.735 : 050
10:49:44.690 : 050
10:50:08.527 : 
10:50:08.832 : 050
10:50:09.782 : 050

**Comment by Jason2866** [comment]
```
00:00:00.032 SER: Set to 7E1 115200 bit/s
00:00:00.032 SER: HWCDC supports 115200 bit/s only
```
I have doubts using really rx/tx. Looks like the config is using cdc serial where the speed and config can not be changed.
